crema/qvalues.py

```diff
@@ -132,3 +132,192 @@ def _fdr2qvalue(scores, fdr):
+    """
+    # TODO try except and some error checking
+
+    num_targets = target_scores.shape[0]
+    num_decoys = decoy_scores.shape[0]
+
+    # calculate p-values from scores
+    nDecoys = 1
+    posSame = 0
+    negSame = 0
+    pValList = []
+    curScore = None
+    for score, target in zip(combined_score, combined_score_target):
+        if target:
+            posSame += 1
+        else:
+            negSame += 1
+
+        # TODO Unclear if need if statement. This appears in Percolator code
+        # however, adding it makes Crema and Percolator pValList different length
+        # if curScore != score:
+        for ix in range(0, posSame):
+            pValList.append(nDecoys + (negSame * (ix + 1)) / (posSame + 1))
+
+        nDecoys += negSame
+        negSame = 0
+        posSame = 0
+        curScore = score
+    pValList = np.array(pValList) / nDecoys
+
+    # calculate pi0
+    pi0 = estimate_pi0(pValList)
+    fdrmod = calculate_mixmax_qval(
+        np.array(target_scores), np.array(decoy_scores), pi0
+    )
+    return (pi0, fdrmod)
+
+
+@nb.njit
+def estimate_pi0(pval_list):
+    """
+    Estimates pi0. Add description. TODO
+
+    Parameters
+    -------
+    pval_list : numpy.ndarray of float
+        A list of p-values sorted from smallest to largest.
+
+    Returns
+    -------
+    pi0 : float
+        Estimated pi_zero.
+    """
+    numLambda = 100
+    maxLambda = 0.5
+    numBoot = 100
+
+    n_pval = pval_list.size
+    lambda_list = []
+    pi0s_list = []
+    for idx in range(0, numLambda):
+        cur_lambda = ((idx + 1) / numLambda) * maxLambda
+
+        # Find the index of the first element in pValList
+        # that is > lambda.
+        start = np.searchsorted(pval_list, cur_lambda)
+        W1 = n_pval - start
+        pi0 = W1 / n_pval / (1.0 - cur_lambda)
+
+        if pi0 > 0.0:
+            lambda_list.append(cur_lambda)
+            pi0s_list.append(pi0)
+
+    assert len(pi0s_list) != 0, (
+        "Error in the input data: "
+        "too good separation between target and decoy PSMs."
+    )
+
+    minPi0 = min(pi0s_list)
+
+    mse_list = np.zeros(len(pi0s_list))
+    max_size = 1000
+    # Examine which lambda level that is most stable under bootstrap
+    for i in range(0, numBoot):
+        # Create an array of bootstrapped p-values, and sort in ascending order.
+        num_draw = min(n_pval, max_size)
+        pBoot_list = np.random.choice(pval_list, size=num_draw, replace=True)
+        pBoot_list.sort()
+
+        for idx in range(0, len(lambda_list)):
+            start = np.searchsorted(pBoot_list, lambda_list[idx])
+            W1 = num_draw - start
+            pi0Boot = W1 / num_draw / (1.0 - lambda_list[idx])
+
+            # Estimated mean-squared error
+            mse_list[idx] += (pi0Boot - minPi0) * (pi0Boot - minPi0)
+
+    # Which index did the iterator go?
+    minIdx = np.argmin(mse_list)
+    pi0 = max(min(pi0s_list[minIdx], 1.0), 0.0)
+    return pi0
```

Comment:
**praise:** Cool! I've never known how pi0 was actually estimated haha.

Reply:
I had to read ALL of assign-confidence to figure it out...it wasn't fun lol